Mathlib/SetTheory/Ordinal/Arithmetic.lean

```diff
@@ -301,6 +301,61 @@ theorem limitRecOn_limit {C} (o H₁ H₂ H₃ h) :
     @limitRecOn C o H₁ H₂ H₃ = H₃ o h fun x _h => @limitRecOn C x H₁ H₂ H₃ := by
   simp_rw [limitRecOn, SuccOrder.prelimitRecOn_limit _ _ h.isSuccPrelimit, dif_neg h.1]
 
+/-- Bounded recursion on ordinals. Similar to `limitRecOn`, with the assumption `o < l`
+  added to all cases. The final term's domain is the ordinals below `Iio l`. -/
+@[elab_as_elim]
+def boundedLimitRec {l : Ordinal} (hLim : l.IsLimit) {C : Iio l → Sort*} (H₁ : C ⟨0, hLim.pos⟩)
+    (H₂ : (o : Iio l) → C o → C ⟨o + 1, hLim.succ_lt o.2⟩)
+    (H₃ : (o : Iio l) → IsLimit o → (Π o' < o, C o') → C o) :
+    (o : Iio l) → C o :=
+  fun o ↦ limitRecOn (C := fun p ↦ (h : p < l) → C ⟨p, h⟩) o.1 (fun _ ↦ H₁)
+    (fun o ih h ↦ H₂ ⟨o, _⟩ <| ih <| (lt_succ o).trans h)
+    (fun _o ho ih _ ↦ H₃ _ ho fun _o' h ↦ ih _ h _) o.2
+
+@[simp]
+theorem boundedLimitRec_zero {l} (lLim : l.IsLimit) {C} (H₁ H₂ H₃) :
+    @boundedLimitRec l lLim C H₁ H₂ H₃ ⟨0, lLim.pos⟩ = H₁ := by
+  simp_all only [boundedLimitRec, limitRecOn_zero, id_eq,
+    eq_mpr_eq_cast, eq_mp_eq_cast, cast_cast, cast_eq]
+
+@[simp]
+theorem boundedLimitRec_succ {l} (lLim : l.IsLimit) {C} (o : Iio l) (H₁ H₂ H₃) :
+    @boundedLimitRec l lLim C H₁ H₂ H₃ ⟨succ o.1, lLim.succ_lt o.2⟩ = H₂ o
+    (@boundedLimitRec l lLim C H₁ H₂ H₃ o) := by
+  simp_all only [add_one_eq_succ, boundedLimitRec, eq_mpr_eq_cast, eq_mp_eq_cast,
+    limitRecOn_succ, dite_true, cast_cast, cast_eq]
+
+@[simp]
+theorem boundedLimitRec_limit {l : Ordinal} (lLim : l.IsLimit) {C} (o H₁ H₂ H₃ oLim) :
```

Comment:
We don't have many `simp` lemmas on specific limit ordinals, so in practice this will rarely if ever fire as a `simp` lemma. `limitRecOn_limit` shouldn't have been a `simp` lemma either.